### PR TITLE
Flatten exceptions in dynamo

### LINKED ISSUE #93571: Rethink exception rethrowing strategy
### 🐛 Describe the bug

After https://github.com/pytorch/pytorch/pull/87440 there are a few things that no longer make sense in the new regime that we should think about changing.

* @mlazos remarked to me that now that we aren't suppressing errors by default, we shouldn't log them to stderr. This makes sense. Essentially, IF we suppress an error, we should log it to stderr; otherwise, we shouldn't.
* Rethrowing exceptions into TorchDynamoInternalError or similar error classes is annoying, because it means by default you get an uninformative stack trace at the bottom (and have to scroll up to find the useful stack trace). We should consider NOT rethrowing. The rethrowing is also why benchmark fails to report accurate exception info.
* It would be useful to know contexts where minification can occur, and if an exception is thrown in such a context, add some extra context saying that "hey, you can run the minifier like this to get a minified repro". This is in tension with bullet two.

cc @soumith @msaroufim @wconstab @ngimel @bdhirsh @Chillee @anijain2305 @jansel

### Error logs

n/a

### Minified repro

n/a

### PR DESCRIPTION
**Pull request by mlazos**
Fixes https://github.com/pytorch/pytorch/issues/93571

[before and after](https://gist.github.com/mlazos/256b0e8f0f98495752a22b960e9f4fcb)

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by ezyang — CHANGES_REQUESTED** [review]
I can't accept this. You've lost the PrimTorch stack trace, which is the most important part for internal debugging purposes!

**Comment by mlazos** [comment]
> I can't accept this. You've lost the PrimTorch stack trace, which is the most important part for internal debugging purposes!

Oh oops, I can apply this rule recursively and get it back. I thought this was perfect ootb!!

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -1314,9 +1314,7 @@ def run_node(tracer, node, args, kwargs, nnmodule):
             assert "example_value" in node.meta
             return node.meta["example_value"]
     except Exception as e:
-        raise RuntimeError(
-            f"Failed running {op} {node.target}(*{args}, **{kwargs}):\n{e}\n(scroll up for backtrace)"
-        ) from e
+        raise RuntimeError(str(e)).with_traceback(e.__traceback__) from None
```
Comment (ezyang) [inline torch/_dynamo/utils.py:1317]: The incremental info (op args kwargs) here was very useful, please preserve it.

**Comment by ezyang** [comment]
I'm not sure how I feel about str'ing the exception. Can we preserve the original structure maybe? IDK.

**Comment by mlazos** [comment]
> I'm not sure how I feel about str'ing the exception. Can we preserve the original structure maybe? IDK.

Yeah I actually ended up keeping the structure in a spot where it was needed in an upstream handler. Do you have a specific need in mind? If anyone needs the `__cause__` to be populated in the future, you're allowed to do new_exc.with_traceback(old_exc) from old_exc to preserve the structure, it's just nobody else is using that rn.

**Comment by ezyang** [comment]
You don't want `from old_exc` because that will trigger chained printing, and if you're overriding the traceback you definitely don't want chaining.

I guess we can give this a try. For downstream consumers who want the original error, this is going to depend a lot on what exactly those systems are doing. Maybe we can reach out to some folks who work on those systems to figure it out.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**